genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/S3ArchivalServiceImpl.java

```diff
@@ -0,0 +1,172 @@
+        @NotNull final File file,
+        @NotNull final URI targetLocationURI
+    ) throws ArchivalException {
+        log.info("Archiving to location: {}", targetLocationURI);
+
+        try {
+            final AmazonS3URI targetLocationAmazonS3URI = new AmazonS3URI(targetLocationURI);
+            archiveRecursively(file, targetLocationAmazonS3URI);
+        } catch (Exception e) {
+            log.info("Error archiving to target location: {} ", targetLocationURI);
+            throw new ArchivalException("Error archiving file: " + file.getName(), e);
+        }
+    }
+
+    /**
+     * Traverse a directory recursively and archive the files and directories
+     * underneath it. If the directory is empty, create an empty object in S3.
+     * For a regular child file create an object in S3 and copy the file contents.
+     * For a empty child directory create an empty object in S3
+     *
+     * @param fileToArchive     file/dir to archive.
+     * @param targetLocationURI target location URI for storing the file in S3.
+     */
+    private void archiveRecursively(
+        @NotNull final File fileToArchive,
+        @NotNull final AmazonS3URI targetLocationURI
+    ) {
+        final File[] children = fileToArchive.listFiles();
+
+        //children == null if fileToArchive is not a directory.
+        //children is a empty array if the directory is empty
+        if (children == null || children.length == 0) {
+            archiveFile(fileToArchive, targetLocationURI);
+        } else { //Found a non empty directory, archive recursively
+            for (File childFile : children) {
+                archiveRecursively(
+                    childFile,
+                    new AmazonS3URI(targetLocationURI + File.separator + childFile.getName())
+                );
+            }
+        }
+    }
+
+    /**
+     * Archive a single file by creating an object in S3. For a file store its contents
+     * For a dir create a object with empty content.
+     *
+     * @param fileToArchive     file/dir to archive
+     * @param targetLocationURI base location for storing the file in S3
+     */
+    private void archiveFile(
+        @NotNull final File fileToArchive,
+        @NotNull final AmazonS3URI targetLocationURI
+    ) {
+        if (!fileToArchive.isDirectory()) {
+            uploadFile(targetLocationURI.getBucket(), targetLocationURI.getKey(), fileToArchive);
+        } else {
+            createEmptyFolder(targetLocationURI.getBucket(), targetLocationURI.getKey());
+        }
+    }
+
+    //Upload a file to the given bucket and key.
+    private void uploadFile(final String bucketName, final String key, final File file) {
+        amazonS3.putObject(new PutObjectRequest(bucketName, key, file));
+    }
+
+    //Create an object representing an empty folder. Key ends with File.separator.
+    //Content of the object is empty
+    private void createEmptyFolder(final String bucketName, final String key) {
```

Comment:
I'm not sure I understand the purpose of this method. S3 has no concept of `folder` so there's no real need to create one. You should just be able to keep uploading files and the prefixes will create "folders".

It seems like maybe you're trying to mimic rsync here so that the entire structure matches exactly. I wonder if that is necessary. I'm not sure if there is value in representing empty directories on a read of this archive location. Just wasted time/API calls. @mprimi thoughts?